### ISSUE TITLE
`leave_workspace` returns `false` the first time, and `true` the second time it is called.

### ISSUE THREAD

**Issue by ghyatzo** [hexedtech/codemp#63]
after calling it the first time `client.active_workspaces()` is empty.
is this plugin specific or do you also incur in this issue?

**Comment by alemidev** [comment]
this is (sadly) expected: while closing resources (leave buffer/workspace/client) we try to consume the Arc we have internally. If this succeeds, it means we had the last reference, so we return `true`. If the Arc isn't the last one, we still end up without a reference but knowing there are others, so we return `false`. Trying to close that resource again will be a no-op since there isn't any other reference to work with, so we just answer `true`. That last result is however a lie if it was `false` the first time and nothing more has been done: there will still be dangling references preventing the resource from being freed!

So basically: this is a badly designed expected behavior. We likely need to rethink how to handle dangling refs across FFI!

**Comment by alemidev** [comment]
an idea to addess this behavior is to make all structs passing ffi just Weak<> wrappers and not full Arc<> wrappers, so they won't hold stuff from being dropped, but implies heavy changes on the api
